common/workunit/workflow.cpp

```diff
@@ -613,12 +808,888 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+    if(totalDependencies == totalDependentSuccessors)
+        LOG(MCworkflow, "dependency and dependent successor count is consistent");
+    else
+        LOG(MCworkflow, "dependency and dependent successor count is inconsistent");
+#endif
+}
+CCloneWorkflowItem * WorkflowMachine::insertLogicalPredecessor(unsigned successorWfid)
+{
+    unsigned wfid = workflow->count() + logicalWorkflow.size()+1;
+#ifdef TRACE_WORKFLOW
+        LOG(MCworkflow, "new predecessor workflow item %u has been created", wfid);
+#endif
+
+    CCloneWorkflowItem * predecessor = new CCloneWorkflowItem(wfid); //initialise the intermediary
+    Owned<IRuntimeWorkflowItem> tmp = predecessor;
+    logicalWorkflow.push_back(tmp); //adding it to the workflow array
+
+    markDependencies(successorWfid, predecessor, false);
+    return predecessor;
+}
+//logicalPredecessor is for Sequential/Condition/Contingency
+void WorkflowMachine::markDependencies(unsigned int wfid, CCloneWorkflowItem *logicalPredecessor, bool prevOrdered)
+{
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "Called mark dependents on item %u", wfid);
+#endif
+
+    CCloneWorkflowItem & item = queryWorkflowItem(wfid);
+    //Ordered causes the effect of logicalPredecessor to skip a generation
+    bool alreadyProcessed = !item.isDependentSuccessorsEmpty();
+    if(!prevOrdered)
+    {
+        if(logicalPredecessor)
+        {
+            if(!alreadyProcessed)
+                logicalPredecessor->addLogicalSuccessor(&item);
+        }
+        else
+        {
+            //if logicalPredecessor is nullptr, then the item should be marked as active, since it doesn't have to follow on from anything else to execute.
+            if(!item.isActive())
+            {
+                item.activate();
+                //This checks that the current item has already performed the markDependencies algorithm
+                if(alreadyProcessed)
+                {
+                    activateDependencies(item);
+                }
+            }
+        }
+    }
+    if(alreadyProcessed)
+        return;
+
+    Owned<IWorkflowDependencyIterator> iter = item.getDependencies();
+    //For Non-Condition items
+    if(item.queryMode() != WFModeCondition)
+    {
+#ifdef TRACE_WORKFLOW
+        LOG(MCworkflow, "Item %u is a non-condition item", wfid);
+#endif
+
+        CCloneWorkflowItem * prev = nullptr;
+        bool thisOrdered = false;
+        for(iter->first(); iter->isValid(); iter->next())
+        {
+            CCloneWorkflowItem & cur = queryWorkflowItem(iter->query());
+            if(prev)
+            {
+                //prev is the logical predecessor to cur.
+                switch(item.queryMode())
+                {
+                case WFModeOrdered:
+                    //this successorship should be added even if cur has already marked dependents
+                    cur.addLogicalSuccessor(prev);
+                    thisOrdered = true;
```

Comment:
worth adding a comment.
//Note: thisOrdered is false for the first ORDERED action.